drivers/iio/frequency/adf4159.c

```
@@ -0,0 +1,1094 @@
+	adf4159_property_bool(st, "adi,txdata-trigger-delay-enable",
+			      &conf->txdata_trig_delay_en);
+
+	return conf;
+}
+
+static unsigned long adf4159_clk_recalc_rate(struct clk_hw *hw,
+		unsigned long parent_rate)
+{
+	unsigned long long rate;
+
+	rate = adf4159_pll_fract_n_get_rate(to_clk_priv(hw)->st);
+
+	return to_ccf_scaled(rate, &to_clk_priv(hw)->scale);
+}
+
+static long adf4159_clk_round_rate(struct clk_hw *hw, unsigned long rate,
+		unsigned long *parent_rate)
+{
+	return rate;
+}
+
+static int adf4159_clk_set_rate(struct clk_hw *hw, unsigned long rate,
+		unsigned long parent_rate)
+{
+	struct adf4159_state *st = to_clk_priv(hw)->st;
+
+	return adf4159_setup(st, parent_rate,
+		from_ccf_scaled(rate, &to_clk_priv(hw)->scale));
+}
+
+static int adf4159_clk_enable(struct clk_hw *hw)
+{
+	to_clk_priv(hw)->enabled = true;
+
+	return 0;
+}
+
+static void adf4159_clk_disable(struct clk_hw *hw)
+{
+	to_clk_priv(hw)->enabled = false;
+}
+
+static int adf4159_clk_is_enabled(struct clk_hw *hw)
+{
+	return to_clk_priv(hw)->enabled;
+}
+
+static const struct clk_ops clkout_ops = {
+	.recalc_rate = adf4159_clk_recalc_rate,
+	.round_rate = adf4159_clk_round_rate,
+	.set_rate = adf4159_clk_set_rate,
+	.enable = adf4159_clk_enable,
+	.disable = adf4159_clk_disable,
+	.is_enabled = adf4159_clk_is_enabled,
+};
+
+static int adf4159_probe(struct spi_device *spi)
+{
+	const struct spi_device_id *id = spi_get_device_id(spi);
+	struct iio_dev *indio_dev;
+	struct adf4159_state *st;
+	int ret;
+
+	indio_dev = devm_iio_device_alloc(&spi->dev, sizeof(*st));
+	if (indio_dev == NULL)
+		return -ENOMEM;
+
+	st = iio_priv(indio_dev);
+
+	st->clk = devm_clk_get(&spi->dev, "clkin");
+	if (IS_ERR(st->clk))
+		return -EPROBE_DEFER;
+
+	ret = clk_prepare_enable(st->clk);
+	if (ret < 0)
+		return ret;
+
+	st->reg = devm_regulator_get(&spi->dev, "vcc");
+	if (!IS_ERR(st->reg)) {
+		ret = regulator_enable(st->reg);
+		if (ret)
+			goto error_disable_clk;
+	}
```

I don't think we can ignore errors here. If we do not define a regulator in dts, a dummy one will given to us... So if we get an error, it might be something that we should not ignore. I would just `if (IS_ERR) return PTR_ERR()` and assume reg is fine after that...

---

yep